File: src/shared/api/dropProxy.js
import { invoke } from '@tauri-apps/api/core';

async function uint8ArrayToNumberArrayChunked(data, chunkSize = 256 * 1024) {
  if (!data) return [];
  if (Array.isArray(data)) return data;
  const u8 = data instanceof Uint8Array ? data : new Uint8Array(data);
  const result = new Array(u8.length);
  for (let offset = 0; offset < u8.length; offset += chunkSize) {
    const end = Math.min(u8.length, offset + chunkSize);
    for (let index = offset; index < end; index += 1) {
      result[index] = u8[index];
    }
    await new Promise((resolve) => setTimeout(resolve, 0));
  }
  return result;
}

export async function ensureDropProxy() {
  return await invoke('drop_proxy_ensure');
}

export async function showDropProxy(bounds) {
  return await invoke('drop_proxy_show', { bounds });
}

export async function hideDropProxy() {
  return await invoke('drop_proxy_hide');
}

export async function disposeDropProxy() {
  return await invoke('drop_proxy_dispose');
}

export async function routeDropProxyPathsAtCursor(paths, cursorPos) {
  return await invoke('drop_proxy_route_paths_at_cursor', { paths, cursorPos });
}

export async function saveDropProxyResource(filename, data) {
  const payloadData = await uint8ArrayToNumberArrayChunked(data);
  return await invoke('drop_proxy_save_resource', {
    payload: {
      filename,
      data: payloadData,
    },
  });
}

export async function saveDropProxyUrl(filename, url) {
  return await invoke('drop_proxy_save_url', {
    payload: {
      filename,
      url,
    },
  });
}

export async function cleanupDropProxyOrphanResources(minAgeMs = 5000) {
  return await invoke('drop_proxy_cleanup_orphan_resources', { minAgeMs });
}

File: src-tauri/src/commands/drop_proxy.rs
use serde::Deserialize;
use tauri::{AppHandle, WebviewWindow};

use crate::windows::drop_proxy::{
    self, DropProxyBounds, DropProxyCleanupPayload, DropProxyCursorPosition, DropProxyResourcePayload,
    DropProxyRouteResult,
};

const DEFAULT_CLEANUP_MIN_AGE_MS: u64 = 5000;

#[tauri::command]
pub async fn drop_proxy_ensure(app: AppHandle) -> Result<(), String> {
    drop_proxy::ensure_drop_proxy(app).await
}

#[tauri::command]
pub async fn drop_proxy_show(window: WebviewWindow, bounds: DropProxyBounds) -> Result<(), String> {
    drop_proxy::show_drop_proxy(window, bounds).await
}

#[tauri::command]
pub fn drop_proxy_hide(app: AppHandle) -> Result<(), String> {
    drop_proxy::hide_drop_proxy(&app)
}

#[tauri::command]
pub fn drop_proxy_dispose(app: AppHandle) -> Result<(), String> {
    drop_proxy::dispose_drop_proxy(&app)
}

#[tauri::command]
pub fn drop_proxy_route_paths_at_cursor(
    app: AppHandle,
    window: WebviewWindow,
    paths: Vec<String>,
    cursor_pos: DropProxyCursorPosition,
) -> Result<DropProxyRouteResult, String> {
    drop_proxy::route_paths_at_cursor(&app, window.label(), paths, cursor_pos)
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DropProxySaveResourceRequest {
    pub filename: String,
    pub data: Vec<u8>,
}

#[tauri::command]
pub async fn drop_proxy_save_resource(payload: DropProxySaveResourceRequest) -> Result<DropProxyResourcePayload, String> {
    drop_proxy::save_drop_resource(payload.filename, payload.data).await
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DropProxySaveUrlRequest {
    pub filename: String,
    pub url: String,
}

#[tauri::command]
pub async fn drop_proxy_save_url(payload: DropProxySaveUrlRequest) -> Result<DropProxyResourcePayload, String> {
    drop_proxy::save_drop_url(payload.filename, payload.url).await
}

#[tauri::command]
pub async fn drop_proxy_cleanup_orphan_resources(min_age_ms: Option<u64>) -> Result<DropProxyCleanupPayload, String> {
    drop_proxy::cleanup_orphan_resources(min_age_ms.unwrap_or(DEFAULT_CLEANUP_MIN_AGE_MS)).await
}
